x-pack/libbeat/management/unit.go

```diff
@@ -0,0 +1,409 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/beats/v7/libbeat/management/status"
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// unitState is the current state of a unit
+type unitState struct {
+	state status.Status
+	msg   string
+}
+
+// agentUnit implements status.StatusReporter and holds an unitState
+// fot the input as well as a unitState for each stream of
+// the input in when this a client.UnitTypeInput.
+type agentUnit struct {
+	softDeleted  bool
+	mtx          sync.Mutex
+	logger       *logp.Logger
+	clientUnit   *client.Unit
+	input        unitState
+	streamIDs    []string
+	streamStates map[string]unitState
+}
+
+// getUnitState converts status.Status to client.UnitState
+func getUnitState(s status.Status) client.UnitState {
+	switch s {
+	case status.Unknown:
+		// must be started if its unknown
+		return client.UnitStateStarting
+	case status.Starting:
+		return client.UnitStateStarting
+	case status.Configuring:
+		return client.UnitStateConfiguring
+	case status.Running:
+		return client.UnitStateHealthy
+	case status.Degraded:
+		return client.UnitStateDegraded
+	case status.Failed:
+		return client.UnitStateFailed
+	case status.Stopping:
+		return client.UnitStateStopping
+	case status.Stopped:
+		return client.UnitStateStopped
+	}
+	// unknown again?
```

Comment:
tbh I don't know 😆 this came as is from [here](https://github.com/elastic/beats/pull/39209/files#diff-dd024b66f2ed585cbbfce1d8a0aeba55183ab0d04bed2a67f7738f62b5658e63L895)

Reply:
The client should have had an unknown value, which should also have been the zero value, but too annoying to change that everywhere now.

If you don't want to create an unknown state, failed is best because it will be noticed.